src/misc/inc/MultiDQuadratureBase.h

```diff
@@ -0,0 +1,78 @@
+//-----------------------------------------------------------------------bl-
+//--------------------------------------------------------------------------
+//
+// QUESO - a library to support the Quantification of Uncertainty
+// for Estimation, Simulation and Optimization
+//
+// Copyright (C) 2008-2017 The PECOS Development Team
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the Version 2.1 GNU Lesser General
+// Public License as published by the Free Software Foundation.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc. 51 Franklin Street, Fifth Floor,
+// Boston, MA  02110-1301  USA
+//
+//-----------------------------------------------------------------------el-
+
+#ifndef UQ_MULTI_D_QUADRATURE_BASE_H
+#define UQ_MULTI_D_QUADRATURE_BASE_H
+
+#include <queso/QuadratureBase.h>
+#include <queso/VectorSubset.h>
+#include <queso/SharedPtr.h>
+
+namespace QUESO
+{
+  // Forward declarations
+  class GslVector;
+  class GslMatrix;
+
+  //! Base class for multi-dimensional quadrature rules
+  /*! Base class for numerical integration of functions of multiple variables
+   *  using quadrature rules. Subclasses will implement the actual quadrature rule
+   *  and should populate the positions and weights. The integration will be performed
+   *  over the user-supplied domain.
+   *
+   *  The intended use of these rules is that the constructor builds the rule and populates
+   *  the positions and weights and then the user can perform the sum \f$ \sum_{i=1}^N f(x_i) w_i \f$.
+   */
+  template <class V = GslVector, class M = GslMatrix>
+  class MultiDQuadratureBase : public QuadratureBase
+  {
+  public:
+
+    MultiDQuadratureBase( const VectorSubset<V,M> & domain )
+      : QuadratureBase(),
+        m_domain(domain)
+    {}
+
+    virtual ~MultiDQuadratureBase(){};
+
+    const std::vector<typename QUESO::SharedPtr<V>::Type> & positions() const
+    { queso_assert(!m_positions.empty());
+      return m_positions; }
+
+    const VectorSubset<V,M> & getDomain() const
+    { return m_domain; }
+
+  protected:
+
+    //! Locations of quadrature points
+    /*! \todo We should make this a vector of std::unique_ptr when we go full C++11. */
```

Comment:
Yeah, and it would make sense here since we never intend to transfer ownership away, so someone would have to explicitly invoke `std::move` if they tried to.